File: frontend/src/modules/admin/ingredients/services/ingredients.service.ts
import { apiClient } from '@/core/config/axios-instance.config'
import type { PaginatedResponse } from '@/core/utils/pagination.utils'
import { buildRequestFilter } from '@/core/utils/request-filters.utils'
import type {
	CreateIngredientCategoryDTO,
	CreateIngredientDTO,
	IngredientCategoryDTO,
	IngredientCategoryFilter,
	IngredientFilter,
	IngredientsDTO,
	IngredientTranslationsDTO,
	UpdateIngredientCategoryDTO,
	UpdateIngredientDTO,
} from '../models/ingredients.model'

class IngredientsService {
	async getIngredients(filter?: IngredientFilter) {
		const response = await apiClient.get<PaginatedResponse<IngredientsDTO[]>>('/ingredients', {
			params: buildRequestFilter(filter),
		})
		return response.data
	}

	async getIngredientById(id: number) {
		const response = await apiClient.get<IngredientsDTO>(`/ingredients/${id}`)
		return response.data
	}

	async createIngredient(dto: CreateIngredientDTO) {
		const response = await apiClient.post<void>('/ingredients', dto)
		return response.data
	}

	async updateIngredient(id: number, dto: UpdateIngredientDTO) {
		const response = await apiClient.put<void>(`/ingredients/${id}`, dto)
		return response.data
	}

	async deleteIngredient(id: number) {
		await apiClient.delete<void>(`/ingredients/${id}`)
	}

	async getIngredientCategories(filter?: IngredientCategoryFilter) {
		const response = await apiClient.get<PaginatedResponse<IngredientCategoryDTO[]>>(
			'/ingredient-categories',
			{ params: buildRequestFilter(filter) },
		)
		return response.data
	}

	async getIngredientCategoryById(id: number) {
		const response = await apiClient.get<IngredientCategoryDTO>(`/ingredient-categories/${id}`)
		return response.data
	}

	async createIngredientCategory(dto: CreateIngredientCategoryDTO) {
		const response = await apiClient.post<void>('/ingredient-categories', dto)
		return response.data
	}

	async updateIngredientCategory(id: number, dto: UpdateIngredientCategoryDTO) {
		const response = await apiClient.put<void>(`/ingredient-categories/${id}`, dto)
		return response.data
	}

	async deleteIngredientCategory(id: number) {
		await apiClient.delete<void>(`/ingredient-categories/${id}`)
	}

	async upsertIngredientTranslations(productId: number, data: IngredientTranslationsDTO) {
		try {
			const response = await apiClient.post<void>(`/ingredients/${productId}/translations`, data)
			return response.data
		} catch (error) {
			console.error('Failed to upsert ingredient translations: ', error)
			throw error
		}
	}

	async getIngredientTranslations(productId: number) {
		try {
			const response = await apiClient.get<IngredientTranslationsDTO>(
				`/ingredients/${productId}/translations`,
			)
			return response.data
		} catch (error) {
			console.error('Failed to get ingredient translations: ', error)
			throw error
		}
	}
}

export const ingredientsService = new IngredientsService()

File: backend/internal/modules/ingredients/ingredients.service.go
package ingredients

import (
	"fmt"

	"github.com/Global-Optima/zeep-web/backend/internal/data"
	"github.com/Global-Optima/zeep-web/backend/internal/modules/ingredients/types"
	"github.com/Global-Optima/zeep-web/backend/pkg/utils"
	"go.uber.org/zap"
)

type IngredientService interface {
	CreateIngredient(dto *types.CreateIngredientDTO) (uint, error)
	UpdateIngredient(ingredientID uint, dto *types.UpdateIngredientDTO) error
	DeleteIngredient(ingredientID uint) error
	GetIngredientByID(ingredientID uint) (*types.IngredientDTO, error)
	GetTranslatedIngredientByID(locale data.LanguageCode, ingredientID uint) (*types.IngredientDTO, error)
	GetIngredientsByIDs(ingredientIDs []uint) ([]types.IngredientDTO, error)
	GetIngredients(locale data.LanguageCode, filter *types.IngredientFilter) ([]types.IngredientDTO, error)

	UpsertIngredientTranslations(id uint, dto *types.IngredientTranslationsDTO) error
}

type ingredientService struct {
	repo               IngredientRepository
	transactionManager TransactionManager
	logger             *zap.SugaredLogger
}

func NewIngredientService(repo IngredientRepository, transactionManager TransactionManager, logger *zap.SugaredLogger) IngredientService {
	return &ingredientService{
		repo:               repo,
		transactionManager: transactionManager,
		logger:             logger,
	}
}

func (s *ingredientService) CreateIngredient(dto *types.CreateIngredientDTO) (uint, error) {
	ingredient, err := types.ConvertToIngredientModel(dto)
	if err != nil {
		return 0, err
	}

	id, err := s.repo.CreateIngredient(ingredient)
	if err != nil {
		wrappedErr := utils.WrapError("Failed to create ingredient:", err)
		s.logger.Error(wrappedErr)
		return 0, err
	}
	return id, nil
}

func (s *ingredientService) UpdateIngredient(ingredientID uint, dto *types.UpdateIngredientDTO) error {
	ingredient, err := s.repo.GetRawIngredientByID(ingredientID)
	if err != nil {
		s.logger.Error("Failed to update ingredient:", err)
		return err
	}

	err = types.ConvertToUpdateIngredientModel(dto, ingredient)
	if err != nil {
		return err
	}

	if err := s.repo.SaveIngredient(ingredientID, ingredient); err != nil {
		s.logger.Error("Failed to update ingredient:", err)
		return err
	}
	return nil
}

func (s *ingredientService) DeleteIngredient(ingredientID uint) error {
	if err := s.repo.DeleteIngredient(ingredientID); err != nil {
		s.logger.Error("Failed to delete ingredient:", err)
		return err
	}
	return nil
}

func (s *ingredientService) GetIngredientByID(ingredientID uint) (*types.IngredientDTO, error) {
	ingredient, err := s.repo.GetIngredientByID(ingredientID)
	if err != nil {
		s.logger.Error("Failed to fetch ingredient by ID:", err)
		return nil, err
	}

	return types.ConvertToIngredientResponseDTO(ingredient), nil
}

func (s *ingredientService) GetTranslatedIngredientByID(locale data.LanguageCode, ingredientID uint) (*types.IngredientDTO, error) {
	ingredient, err := s.repo.GetTranslatedIngredientByID(locale, ingredientID)
	if err != nil {
		s.logger.Error("Failed to fetch translated ingredient by ID:", err)
		return nil, err
	}
	return types.ConvertToIngredientResponseDTO(ingredient), nil
}

func (s *ingredientService) GetIngredientsByIDs(ingredientIDs []uint) ([]types.IngredientDTO, error) {
	ingredients, err := s.repo.GetIngredientsWithDetailsByIDs(ingredientIDs)
	if err != nil {
		s.logger.Error("Failed to fetch ingredient by ID:", err)
		return nil, err
	}

	dtos := make([]types.IngredientDTO, len(ingredients))
	for i, ingredient := range ingredients {
		dtos[i] = *types.ConvertToIngredientResponseDTO(&ingredient)
	}

	return dtos, nil
}

func (s *ingredientService) GetIngredients(locale data.LanguageCode, filter *types.IngredientFilter) ([]types.IngredientDTO, error) {
	ingredients, err := s.repo.GetIngredients(locale, filter)
	if err != nil {
		s.logger.Error("Failed to fetch ingredients:", err)
		return nil, err
	}

	return types.ConvertToIngredientResponseDTOs(ingredients), nil
}

func (s *ingredientService) UpsertIngredientTranslations(id uint, dto *types.IngredientTranslationsDTO) error {
	if dto == nil {
		return fmt.Errorf("translations DTO is nil")
	}

	if err := s.transactionManager.UpsertIngredientTranslations(id, dto); err != nil {
		wrappedErr := fmt.Errorf("failed to upsert ingredient translations: %w", err)
		s.logger.Error(wrappedErr)
		return wrappedErr
	}

	return nil
}
